fix: support negative drawing offsets

RaphiaRa opened the pull request:
Tiled's drawing offset was previously cast to u32
and added to the atlas margin, causing crashes when tilesets used negative offsets.

Negative drawing offsets can be useful, for example, to simulate alternative anchors (since Tiled's tile anchors are always bottom left)

This change removes the tileset's drawing offset from the atlas margin and instead applies it as a Transform to the tilemap_entity.

RaphiaRa committed:
fix: support negative drawing offsets

Tiled's drawing offset was previously cast to u32
and added to the atlas margin, causing crashes when tilesets
used negative offsets.

Negative drawing offsets can be useful, for example, to simulate
alternative anchors (since Tiled's tile anchors are always bottom left)

Changes:
- Remove tileset drawing offset from atlas margin and store it in TiledMapTileset
- Add a child draw entity to each tile object entity to decouple logic from rendering
- Apply the drawing offset as a `Transform` to tilemap entities and tile object draw entities

adrien-bon commented:
Had a better look at your PR and the code around it.
I'm not sure the original code was working at all:

- we only considered the "drawing offset" for tilesets using a single image + an atlas. We did not for tileset using an images collection
- even for these tilesets, I struggle to see any visual difference when playing with this offset (eventhough I do reproduce the crash you mention when using a negative offset).

Your fix is more than needed if we want to properly support this feature! :)

After playing a bit with your branch, it seems much better than the current code: no more crash when using a negative offset and there is an actual difference when changing its value, whatever the kind of tileset.

However, I see some differences between the preview in Tiled and what is actually displayed when loading the map with Bevy:

- colliders created from tiles don't have the proper position
- colliders created from tile objects don't have the proper position
- tile objects position seems to be off when they are scaled up / down

You can see this by running the physics example (for instance `physics_avian_orientation`) and manually tweak the tileset offset. Especially visible with the `finite.tmx` map.

Here are some changes I propose to fix that:

- for tiles, you did the right thing: which is to apply the offset on the tilemap entity. However, physics plugin uses the layer entity as reference to setup colliders, so it currently does not account for the tileset offset.
 You could retrieve this offset in the [`get_tiles`](https://github.com/adrien-bon/bevy_ecs_tiled/blob/main/src/physics/collider.rs#L48) function and add it to the tile position it returns.
 I *think* that would do the trick: the collider from this particular tile will have the proper offset.
- for objects, it's a bit more complicated. I agree with you, I'd prefer not to change the actual object position and using a dedicated entity for the actual `Sprite` is a good idea.
 However, we need to somehow pass down the tileset offset to the physics plugin so it can be taken into account.
 We also need to take into account the tile object scaling and apply it to the offset.
  - to apply scaling, you can have a look at what's done [here](https://github.com/adrien-bon/bevy_ecs_tiled/blob/main/src/physics/collider.rs#L124) and basically do the same in the object spawning code to adjust the offset
  - to pass down the offset, you could query for the `TiledObject` child in the physics plugin and retrieve its `Transform` to deduce the offset to apply on the collider ? That feel a bit complicated but should work. I'm all hear if you find an easier solution.

Hopefully it's clear, let me know if you have question or better ideas! :)

Thanks

RaphiaRa commented on src/tiled/map/spawn.rs:
```diff
@@ -419,12 +424,17 @@ fn spawn_objects_layer(
         // Handle objects containing tile data:
         // we want to add a Sprite component to the object entity
         // and possibly an animation component if the tile is animated.
+        let draw_entity = commands.spawn(ChildOf(object_entity)).id();
```
ah yes, I missed that. Adjusted it.

RaphiaRa commented on src/tiled/map/loader.rs:
```diff
@@ -344,10 +344,7 @@ fn tileset_to_tiled_map_tileset(
-                                tileset.offset_x as u32 + tileset.margin,
-                                tileset.offset_y as u32 + tileset.margin,
-                            )),
+                            Some(UVec2::new(tileset.margin, tileset.margin)),
```
adjusted it

RaphiaRa committed:
fix: apply drawing offset to collider positions

- Add drawing offset to tile colliders
- Add drawing offset to tile object colliders

RaphiaRa committed:
fix: scale drawing offset with tile object scaling

Tile objects can be scaled.
This change ensures the drawing offset is adjusted accordingly.

RaphiaRa committed:
chore: remove unused drawing_offset

We now access the tileset directly, so storing
the drawing offset is no longer necessary.

adrien-bon reviewed (commented):
I did not realize we could directly access the tileset offset from the physics plugin.
It feels much better to do that rather than query the object child.

Seems like it works, thanks for the nice work :)
Could you please just take into account my last review about the `Name` component and also update the changelog ?

I'll do some more testings but I think we are close to merge and release after that :)

RaphiaRa committed:
chore: add Name component to visual child entities

RaphiaRa committed:
chore: update changelog

RaphiaRa commented on src/tiled/map/spawn.rs:
```diff
@@ -420,11 +425,11 @@ fn spawn_objects_layer(
-            (Some(sprite), None) => {
-                commands.entity(object_entity).insert(sprite);
+            (Some((sprite, offset_transform)), None) => {
+                commands.spawn((ChildOf(object_entity), sprite, offset_transform));
```
Any conventions for that? I simply called it `TileVisual` for now, or do you want the the object name to also be a part of it
